File: frontend/src/types/topic.ts
export interface TopicModuleInfo {
  id: number
  title: string
}

export interface TopicProgressInfo {
  is_completed: boolean
  time_spent_seconds: number
  first_visited_at: string | null
  completed_at: string | null
}

export interface Topic {
  id: number
  title: string
  content_markdown: string
  video_url: string | null
  estimated_minutes: number
  has_quiz: boolean
  order_index: number
  module: TopicModuleInfo
  progress_info: TopicProgressInfo | null
}

export interface TopicVisitResponse {
  message: string
}

export interface TopicCompleteResponse {
  message: string
  is_completed: boolean
}

export interface TopicTimeResponse {
  message: string
  total_seconds: number
}

File: backend/app/schemas/topic.py
from datetime import datetime
from pydantic import BaseModel


class TopicModuleInfo(BaseModel):
    id: int
    title: str

    model_config = {"from_attributes": True}


class TopicProgressInfo(BaseModel):
    is_completed: bool = False
    time_spent_seconds: int = 0
    first_visited_at: datetime | None = None
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class TopicResponse(BaseModel):
    id: int
    title: str
    content_markdown: str
    video_url: str | None = None
    estimated_minutes: int
    has_quiz: bool
    order_index: int
    module: TopicModuleInfo
    progress_info: TopicProgressInfo | None = None

    model_config = {"from_attributes": True}


class TopicVisitResponse(BaseModel):
    message: str


class TopicCompleteResponse(BaseModel):
    message: str
    is_completed: bool


class TopicTimeRequest(BaseModel):
    seconds: int


class TopicTimeResponse(BaseModel):
    message: str
    total_seconds: int
